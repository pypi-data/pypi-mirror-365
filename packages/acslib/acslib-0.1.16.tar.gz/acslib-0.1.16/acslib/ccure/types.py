from enum import Enum


class ObjectType(Enum):
    CLEARANCE = "clearance"
    CLEARANCE_ASSIGNMENT = "clearance assignment"
    CREDENTIAL = "credential"
    CLEARANCE_ITEM = "clearance item"
    DOOR = "door"
    ISTAR_DOOR = "istar door"
    ELEVATOR = "elevator"
    IMAGE = "image"
    ISTAR_CONTROLLER = "istar controller"
    PERSONNEL = "personnel"
    TIME_SPEC = "time spec"
    GROUP = "group"
    GROUP_MEMBER = "group member"

    @property
    def complete(self):
        if self == self.CLEARANCE:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Clearance"
        if self == self.CLEARANCE_ASSIGNMENT:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.PersonnelClearancePair"
        if self == self.CREDENTIAL:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Credential"
        if self == self.CLEARANCE_ITEM:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.ClearanceItem"
        if self == self.DOOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Door"
        if self == self.ISTAR_DOOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.iStarDoor"
        if self == self.ELEVATOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Elevator"
        if self == self.IMAGE:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Images"
        if self == self.ISTAR_CONTROLLER:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.iStarController"
        if self == self.PERSONNEL:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Personnel"
        if self == self.TIME_SPEC:
            return "SoftwareHouse.CrossFire.Common.Objects.TimeSpec"
        if self == self.GROUP:
            return "SoftwareHouse.CrossFire.Common.Objects.Group"
        if self == self.GROUP_MEMBER:
            return "SoftwareHouse.CrossFire.Common.Objects.GroupMember"


class ImageType(Enum):
    UNKNOWN = 0
    PORTRAIT = 1
    SIGNATURE = 2
    FINGERPRINT = 3
    HANDPRINT = 4
    DYNAMIC_BADGE_IMAGE = 5
    STATIC_BADGE_IMAGE = 6
    SYSTEM_IMAGE = 7
    PRIVATE_IMAGE = 8
    SHARED_IMAGE = 9
