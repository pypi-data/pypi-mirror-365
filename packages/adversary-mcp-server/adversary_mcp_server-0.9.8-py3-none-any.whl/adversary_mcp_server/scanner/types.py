"""Core data types for security vulnerability detection."""

import uuid
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path


class Severity(str, Enum):
    """Security vulnerability severity levels."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Category(str, Enum):
    """Security vulnerability categories."""

    INJECTION = "injection"
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    CRYPTO = "crypto"
    CRYPTOGRAPHY = "cryptography"
    CONFIGURATION = "configuration"
    VALIDATION = "validation"
    LOGGING = "logging"
    DESERIALIZATION = "deserialization"
    SSRF = "ssrf"
    XSS = "xss"
    IDOR = "idor"
    RCE = "rce"
    LFI = "lfi"
    DISCLOSURE = "disclosure"
    ACCESS_CONTROL = "access_control"
    TYPE_SAFETY = "type_safety"
    SECRETS = "secrets"
    DOS = "dos"
    CSRF = "csrf"
    PATH_TRAVERSAL = "path_traversal"
    REDIRECT = "redirect"
    HEADERS = "headers"
    SESSION = "session"
    FILE_UPLOAD = "file_upload"
    XXE = "xxe"
    CLICKJACKING = "clickjacking"
    MISC = "misc"  # Generic category for miscellaneous threats


class Language(str, Enum):
    """Supported programming languages."""

    PYTHON = "python"
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    HTML = "html"
    CSS = "css"
    JSON = "json"
    YAML = "yaml"
    XML = "xml"
    PHP = "php"
    SHELL = "shell"
    DOCKERFILE = "dockerfile"
    GO = "go"
    RUBY = "ruby"
    JAVA = "java"
    CSHARP = "csharp"
    SQL = "sql"
    TERRAFORM = "terraform"
    GENERIC = "generic"  # For unknown file types


class LanguageSupport:
    """Centralized language support configuration.

    This class contains all language-related mappings and configurations
    in one place to avoid duplication across modules.
    """

    # Central configuration: Language -> (primary_extension, [all_extensions])
    _LANGUAGE_CONFIG = {
        Language.PYTHON: (".py", [".py", ".pyw", ".pyx"]),
        Language.JAVASCRIPT: (".js", [".js", ".jsx", ".mjs", ".cjs"]),
        Language.TYPESCRIPT: (".ts", [".ts", ".tsx", ".mts", ".cts"]),
        Language.HTML: (
            ".html",
            [
                ".html",
                ".htm",
                ".ejs",
                ".handlebars",
                ".hbs",
                ".mustache",
                ".vue",
                ".svelte",
            ],
        ),
        Language.CSS: (".css", [".css", ".scss", ".sass", ".less", ".stylus"]),
        Language.JSON: (".json", [".json", ".jsonc", ".json5"]),
        Language.YAML: (".yaml", [".yaml", ".yml"]),
        Language.XML: (".xml", [".xml", ".xsl", ".xslt", ".svg", ".rss", ".atom"]),
        Language.PHP: (
            ".php",
            [".php", ".phtml", ".php3", ".php4", ".php5", ".php7", ".phps"],
        ),
        Language.SHELL: (".sh", [".sh", ".bash", ".zsh", ".fish", ".ksh", ".csh"]),
        Language.DOCKERFILE: (".dockerfile", [".dockerfile", "Dockerfile"]),
        Language.GO: (".go", [".go"]),
        Language.RUBY: (".rb", [".rb", ".rbw", ".rake", ".gemspec"]),
        Language.JAVA: (".java", [".java", ".groovy", ".gradle"]),
        Language.CSHARP: (".cs", [".cs", ".csx", ".vb"]),
        Language.SQL: (".sql", [".sql", ".psql", ".mysql", ".sqlite", ".plsql"]),
        Language.TERRAFORM: (".tf", [".tf", ".tfvars", ".hcl"]),
        Language.GENERIC: (
            ".txt",
            [
                ".txt",
                ".md",
                ".cfg",
                ".conf",
                ".ini",
                ".env",
                ".toml",
                ".log",
                ".properties",
            ],
        ),
    }

    @classmethod
    def get_supported_languages(cls) -> list[Language]:
        """Get list of all supported languages."""
        return list(cls._LANGUAGE_CONFIG.keys())

    @classmethod
    def get_extension_to_language_map(cls) -> dict[str, Language]:
        """Get mapping from file extensions to languages."""
        extension_map = {}
        for language, (_, extensions) in cls._LANGUAGE_CONFIG.items():
            for ext in extensions:
                # Handle special case for Dockerfile without extension
                if ext == "Dockerfile":
                    extension_map["dockerfile"] = language  # lowercase for matching
                else:
                    extension_map[ext] = language
        return extension_map

    @classmethod
    def get_language_to_extension_map(cls) -> dict[Language, str]:
        """Get mapping from languages to their primary file extensions."""
        return {
            language: primary_ext
            for language, (primary_ext, _) in cls._LANGUAGE_CONFIG.items()
        }

    @classmethod
    def detect_language(cls, file_path: str | Path) -> Language:
        """Detect language from file path."""
        path = Path(file_path)

        # Handle special case for Dockerfile
        if path.name.lower() in ["dockerfile", "containerfile"]:
            return Language.DOCKERFILE

        extension = path.suffix.lower()
        extension_map = cls.get_extension_to_language_map()

        return extension_map.get(extension, Language.GENERIC)

    @classmethod
    def get_extensions_for_language(cls, language: Language) -> list[str]:
        """Get all file extensions for a specific language."""
        if language in cls._LANGUAGE_CONFIG:
            return cls._LANGUAGE_CONFIG[language][1].copy()
        return []

    @classmethod
    def get_primary_extension(cls, language: Language) -> str:
        """Get the primary file extension for a language."""
        if language in cls._LANGUAGE_CONFIG:
            return cls._LANGUAGE_CONFIG[language][0]
        return ".txt"

    @classmethod
    def get_language_enum_values(cls) -> list[str]:
        """Get list of language enum values for API schemas."""
        return [lang.value for lang in cls.get_supported_languages()]


@dataclass
class ThreatMatch:
    """A detected security threat."""

    # Required fields
    rule_id: str
    rule_name: str
    description: str
    category: Category
    severity: Severity
    file_path: str
    line_number: int

    # Optional fields with defaults
    uuid: str = field(default_factory=lambda: str(uuid.uuid4()))
    column_number: int = 0
    code_snippet: str = ""
    function_name: str | None = None
    exploit_examples: list[str] = field(default_factory=list)
    remediation: str = ""
    references: list[str] = field(default_factory=list)
    cwe_id: str | None = None
    owasp_category: str | None = None
    confidence: float = 1.0  # 0.0 to 1.0
    source: str = "rules"  # Scanner source: "rules", "semgrep", "llm"
    is_false_positive: bool = False  # False positive tracking

    def get_fingerprint(self) -> str:
        """Generate a unique fingerprint for this finding.

        Used to identify the same logical finding across multiple scans
        to preserve UUIDs and false positive markings.

        Returns:
            Unique fingerprint string based on rule_id, file_path, and line_number
        """
        from pathlib import Path

        # Normalize file path to handle relative vs absolute paths
        normalized_path = str(Path(self.file_path).resolve())
        return f"{self.rule_id}:{normalized_path}:{self.line_number}"
