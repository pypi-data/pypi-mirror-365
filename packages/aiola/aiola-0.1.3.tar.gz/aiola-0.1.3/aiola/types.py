from __future__ import annotations

import enum
from collections.abc import Mapping
from dataclasses import dataclass
from typing import IO, TypedDict, Union

from .constants import DEFAULT_AUTH_BASE_URL, DEFAULT_BASE_URL, DEFAULT_HTTP_TIMEOUT, DEFAULT_WORKFLOW_ID


@dataclass
class AiolaClientOptions:
    """Configuration options for Aiola clients."""

    base_url: str | None = DEFAULT_BASE_URL
    auth_base_url: str | None = DEFAULT_AUTH_BASE_URL
    api_key: str | None = None
    access_token: str | None = None
    workflow_id: str = DEFAULT_WORKFLOW_ID
    timeout: float | None = DEFAULT_HTTP_TIMEOUT

    def __post_init__(self) -> None:
        """Validate options after initialization."""
        if not self.api_key and not self.access_token:
            raise ValueError("Either api_key or access_token must be provided")

        if self.api_key is not None and not isinstance(self.api_key, str):
            raise TypeError("API key must be a string")

        if self.access_token is not None and not isinstance(self.access_token, str):
            raise TypeError("Access token must be a string")

        if self.base_url is not None and not isinstance(self.base_url, str):
            raise TypeError("Base URL must be a string")

        if self.auth_base_url is not None and not isinstance(self.auth_base_url, str):
            raise TypeError("Auth base URL must be a string")

        if not isinstance(self.workflow_id, str):
            raise TypeError("Workflow ID must be a string")

        if self.timeout is not None and not isinstance(self.timeout, (int | float)):
            raise TypeError("Timeout must be a number")


class LiveEvents(str, enum.Enum):
    Transcript = "transcript"
    Translation = "translation"
    SentimentAnalysis = "sentiment_analysis"
    Summarization = "summarization"
    TopicDetection = "topic_detection"
    ContentModeration = "content_moderation"
    AutoChapters = "auto_chapters"
    FormFilling = "form_filling"
    EntityDetection = "entity_detection"
    EntityDetectionFromList = "entity_detection_from_list"
    KeyPhrases = "key_phrases"
    PiiRedaction = "pii_redaction"
    Error = "error"
    Disconnect = "disconnect"
    Connect = "connect"


class Segment(TypedDict):
    start: float
    end: float
    text: str


class TranscriptionMetadata(TypedDict):
    """Metadata for transcription results."""

    duration: float
    language: str
    sample_rate: int
    num_channels: int
    timestamp_utc: str
    model_version: str


class TranscriptionResponse(TypedDict):
    """Response from file transcription API."""

    transcript: str
    itn_transcript: str
    segments: list[Segment]
    metadata: TranscriptionMetadata


class SessionCloseResponse(TypedDict):
    """Response from session close API."""

    status: str
    deletedAt: str


class GrantTokenResponse(TypedDict):
    """Response from grant token API."""

    accessToken: str
    sessionId: str


class TranslationPayload(TypedDict):
    src_lang_code: str
    dst_lang_code: str


class EntityDetectionFromListPayload(TypedDict):
    entity_list: list[str]


class _EmptyPayload(TypedDict):
    pass


EntityDetectionPayload = _EmptyPayload
KeyPhrasesPayload = _EmptyPayload
PiiRedactionPayload = _EmptyPayload
SentimentAnalysisPayload = _EmptyPayload
SummarizationPayload = _EmptyPayload
TopicDetectionPayload = _EmptyPayload
ContentModerationPayload = _EmptyPayload
AutoChaptersPayload = _EmptyPayload
FormFillingPayload = _EmptyPayload


class TasksConfig(TypedDict, total=False):
    FORM_FILLING: FormFillingPayload
    TRANSLATION: TranslationPayload
    ENTITY_DETECTION: EntityDetectionPayload
    ENTITY_DETECTION_FROM_LIST: EntityDetectionFromListPayload
    KEY_PHRASES: KeyPhrasesPayload
    PII_REDACTION: PiiRedactionPayload
    SENTIMENT_ANALYSIS: SentimentAnalysisPayload
    SUMMARIZATION: SummarizationPayload
    TOPIC_DETECTION: TopicDetectionPayload
    CONTENT_MODERATION: ContentModerationPayload
    AUTO_CHAPTERS: AutoChaptersPayload


FileContent = Union[IO[bytes], bytes, str]
File = Union[
    # file (or bytes)
    FileContent,
    # (filename, file (or bytes))
    tuple[str | None, FileContent],
    # (filename, file (or bytes), content_type)
    tuple[str | None, FileContent, str | None],
    # (filename, file (or bytes), content_type, headers)
    tuple[
        str | None,
        FileContent,
        str | None,
        Mapping[str, str],
    ],
]
