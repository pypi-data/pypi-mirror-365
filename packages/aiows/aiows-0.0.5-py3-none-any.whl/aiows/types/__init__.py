"""
Type definitions for aiows
"""

from datetime import datetime
from typing import Literal, Tuple
from pydantic import BaseModel, Field


class BaseMessage(BaseModel):
    """Base class for all WebSocket messages"""
    type: str
    timestamp: datetime = Field(default_factory=datetime.now)
    
    def dict(self, **kwargs):
        """Serialize message to dictionary"""
        return super().model_dump(**kwargs)


class ChatMessage(BaseMessage):
    """Chat message with text content"""
    type: Literal["chat"] = "chat"
    text: str
    user_id: int


class JoinRoomMessage(BaseMessage):
    """Message for joining a room"""
    type: Literal["join_room"] = "join_room"
    room_id: str
    user_name: str


class GameActionMessage(BaseMessage):
    """Game action message with coordinates"""
    type: Literal["game_action"] = "game_action"
    action: str
    coordinates: Tuple[int, int] 