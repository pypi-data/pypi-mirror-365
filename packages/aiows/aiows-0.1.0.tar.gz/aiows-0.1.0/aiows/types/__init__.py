"""
Type definitions for aiows
"""

from datetime import datetime
from typing import Literal, Tuple
from pydantic import BaseModel, Field, field_validator

from ..validators import (
    validate_safe_text, validate_username, validate_room_id, 
    validate_game_action, validate_user_id, validate_coordinates
)


class BaseMessage(BaseModel):
    """Base class for all WebSocket messages"""
    type: str
    timestamp: datetime = Field(default_factory=datetime.now)
    
    def dict(self, **kwargs):
        """Serialize message to dictionary"""
        return super().model_dump(**kwargs)


class ChatMessage(BaseMessage):
    """Chat message with text content"""
    type: Literal["chat"] = "chat"
    text: str
    user_id: int
    
    @field_validator('text')
    @classmethod
    def validate_text_field(cls, v):
        return validate_safe_text(cls, v)
    
    @field_validator('user_id', mode='before')
    @classmethod
    def validate_user_id_field(cls, v):
        return validate_user_id(cls, v)


class JoinRoomMessage(BaseMessage):
    """Message for joining a room"""
    type: Literal["join_room"] = "join_room"
    room_id: str
    user_name: str
    
    @field_validator('room_id')
    @classmethod
    def validate_room_id_field(cls, v):
        return validate_room_id(cls, v)
    
    @field_validator('user_name')
    @classmethod
    def validate_user_name_field(cls, v):
        return validate_username(cls, v)


class GameActionMessage(BaseMessage):
    """Game action message with coordinates"""
    type: Literal["game_action"] = "game_action"
    action: str
    coordinates: Tuple[int, int]
    
    @field_validator('action')
    @classmethod
    def validate_action_field(cls, v):
        return validate_game_action(cls, v)
    
    @field_validator('coordinates')
    @classmethod
    def validate_coordinates_field(cls, v):
        return validate_coordinates(cls, v)