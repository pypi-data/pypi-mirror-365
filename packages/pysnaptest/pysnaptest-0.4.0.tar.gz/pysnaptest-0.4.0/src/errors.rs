use std::env::VarError;
use std::fmt::{self, Display, Formatter};

use pyo3::exceptions::PyValueError;
use pyo3::PyErr;
use pythonize::PythonizeError;

#[derive(Debug)]
pub enum PytestInfoError {
    CouldNotSplit(String),
    InvalidEnvVar(VarError),
    NoTestFile,
}

impl Display for PytestInfoError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            PytestInfoError::CouldNotSplit(s) => {
                write!(f, "Expected '::' to be in PYTEST_CURRENT_TEST string ({s})")
            }
            PytestInfoError::InvalidEnvVar(e) => match e {
                VarError::NotPresent => write!(f, "PYTEST_CURRENT_TEST is not set"),
                VarError::NotUnicode(os_string) => {
                    write!(
                        f,
                        "PYTEST_CURRENT_TEST is not a valid unicode string: {os_string:#?}"
                    )
                }
            },
            PytestInfoError::NoTestFile => write!(f, "No test file found"),
        }
    }
}

impl From<PytestInfoError> for PyErr {
    fn from(value: PytestInfoError) -> Self {
        match value {
            PytestInfoError::CouldNotSplit(s) => PyValueError::new_err(format!(
                "Expected '::' to be in PYTEST_CURRENT_TEST string ({s})"
            )),
            PytestInfoError::InvalidEnvVar(ve) => match ve {
                VarError::NotPresent => PyValueError::new_err("PYTEST_CURRENT_TEST is not set"),
                VarError::NotUnicode(os_string) => PyValueError::new_err(format!(
                    "PYTEST_CURRENT_TEST is not a valid unicode string: {os_string:#?}"
                )),
            },
            PytestInfoError::NoTestFile => PyValueError::new_err("No test file found"),
        }
    }
}

#[derive(Debug)]
pub enum SnapError {
    Msg(String),
    Io(std::io::Error),
    Json(serde_json::Error),
    Py(PyErr),
    PytestInfo(PytestInfoError),
    Pythonize(PythonizeError),
    Other(Box<dyn std::error::Error>),
}

impl Display for SnapError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            SnapError::Msg(m) => write!(f, "{m}"),
            SnapError::Io(e) => write!(f, "{e}"),
            SnapError::Json(e) => write!(f, "{e}"),
            SnapError::Py(e) => write!(f, "{e}"),
            SnapError::PytestInfo(e) => write!(f, "{e}"),
            SnapError::Pythonize(e) => write!(f, "{e}"),
            SnapError::Other(e) => write!(f, "{e}"),
        }
    }
}

impl std::error::Error for SnapError {}

impl From<String> for SnapError {
    fn from(value: String) -> Self {
        SnapError::Msg(value)
    }
}

impl From<&str> for SnapError {
    fn from(value: &str) -> Self {
        SnapError::Msg(value.to_string())
    }
}

impl From<std::io::Error> for SnapError {
    fn from(value: std::io::Error) -> Self {
        SnapError::Io(value)
    }
}

impl From<PythonizeError> for SnapError {
    fn from(value: PythonizeError) -> Self {
        SnapError::Pythonize(value)
    }
}

impl From<Box<dyn std::error::Error>> for SnapError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        SnapError::Other(err)
    }
}

impl From<serde_json::Error> for SnapError {
    fn from(value: serde_json::Error) -> Self {
        SnapError::Json(value)
    }
}

impl From<PyErr> for SnapError {
    fn from(value: PyErr) -> Self {
        SnapError::Py(value)
    }
}

impl From<PytestInfoError> for SnapError {
    fn from(value: PytestInfoError) -> Self {
        SnapError::PytestInfo(value)
    }
}

pub type SnapResult<T> = Result<T, SnapError>;
