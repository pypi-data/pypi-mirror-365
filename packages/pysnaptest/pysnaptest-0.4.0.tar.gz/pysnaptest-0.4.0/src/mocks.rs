use std::collections::HashMap;

use insta::assert_json_snapshot as assert_json_snapshot_macro;
use insta::internals::SnapshotContents;
use insta::Snapshot;
use pyo3::prelude::*;
use pyo3::types::{PyDict, PyTuple};

use crate::errors::{SnapError, SnapResult};
use crate::{RedactionType, SnapshotInfo};

macro_rules! snapshot_fn_auto {
    ($f:expr $(, $arg:ident )* ; serialize_macro = $serialize_macro:ident ; result_from_str=$result_from_str:expr) => {{
        let f = $f;
        let name = stringify!($f);
        let module_path = module_path!();

        move |$( $arg ),+, info: &SnapshotInfo, redactions: Option<HashMap<String, RedactionType>>, record: bool| -> SnapResult<_> {
            let finfo = SnapshotInfo {
                snapshot_name: format!("{}_{}", info.snapshot_name, name),
                ..info.clone()
            };
            let snapshot_path = finfo.next_snapshot_path(Some(module_path.to_string()))?;
            let snapshot_name = finfo.snapshot_name();
            let mut settings: insta::Settings = (&finfo).try_into()?;

            for (selector, redaction) in redactions.unwrap_or_default() {
                settings.add_redaction(selector.as_str(), redaction);
            }

            // Serialize the input using the passed closure
            settings.bind(|| {
                $serialize_macro!(format!("{snapshot_name}-request"), ($( $arg ),+));
            });


            if record || !snapshot_path.exists() {
                let result = f($( $arg ),+)?;
                settings.bind(|| {
                    $serialize_macro!(snapshot_name, result);
                });
                Ok(result)
            } else {
                match Snapshot::from_file(&snapshot_path)
                    .map_err(SnapError::from)?
                    .contents()
                {
                    SnapshotContents::Text(content) => {
                        Ok(($result_from_str)(content.to_string())?)
                    },
                    SnapshotContents::Binary(_) => Err(SnapError::from(
                        format!(
                            "Snapshot at {:?} is binary, which is not supported for deserialization",
                            snapshot_path
                        ),
                    )),
                }
            }
        }
    }};
}

macro_rules! snapshot_fn_auto_json {
    ($f:expr $(, $arg:ident )* ; serialize_macro = $serialize_macro:ident ; result_from_str=$result_from_str:expr) => {
        snapshot_fn_auto!($f $(, $arg )* ; serialize_macro = $serialize_macro ; result_from_str=$result_from_str)
    };

    ($f:expr $(, $arg:ident )* ) => {
        snapshot_fn_auto_json!(
            $f,
            $( $arg ),+;
            serialize_macro=assert_json_snapshot_macro;
            result_from_str=|content: String| serde_json::from_str(&content)
        )
    };
}

macro_rules! assert_json_snapshot_depythonize {
    ($snapshot_name:expr, ($arg:expr, $kwargs:expr ) ) => {{
        // Create a tuple of depythonized values

        let rust_args = pythonize::depythonize::<serde_json::Value>($arg as &Bound<PyAny>)
            .expect(&format!("Failed to depythonize args {:?}", $arg));
        let rust_kwargs = Option::<&Bound<'_, PyDict>>::map($kwargs, |kw| {
            pythonize::depythonize::<serde_json::Value>(kw as &Bound<PyAny>)
                .expect(&format!("Failed to depythonize kwargs {:?}", kw))
        });
        let input_json = serde_json::json!({
            "args": rust_args,
            "kwargs": rust_kwargs.unwrap_or(serde_json::Value::Null)
        });

        assert_json_snapshot_macro!($snapshot_name, input_json);
    }};
    ($snapshot_name:expr, $arg:expr) => {{
        Python::with_gil(|py| {
            let bound: &pyo3::Bound<PyAny> = $arg.bind(py);
            let input_tuple = pythonize::depythonize::<serde_json::Value>(&bound)
                .expect(&format!("Failed to depythonize {:?}", $arg));
            assert_json_snapshot_macro!($snapshot_name, input_tuple);
        });
    }};
}

#[pyclass]
#[allow(clippy::type_complexity)]
pub struct PyMockWrapper {
    pub f: Box<
        dyn for<'a> Fn(
                &'a Bound<'_, PyTuple>,
                Option<&'a Bound<'_, PyDict>>,
                &'a SnapshotInfo,
                Option<HashMap<String, RedactionType>>,
                bool,
            ) -> SnapResult<Py<PyAny>>
            + Send
            + Sync,
    >,
    pub snapshot_info: SnapshotInfo,
    pub record: bool,
    pub redactions: Option<HashMap<String, RedactionType>>,
}

#[pymethods]
impl PyMockWrapper {
    #[pyo3(signature = (*args, **kwargs))]
    fn __call__(
        &self,
        args: &Bound<'_, PyTuple>,
        kwargs: Option<&Bound<'_, PyDict>>,
    ) -> PyResult<PyObject> {
        (self.f)(
            args,
            kwargs,
            &self.snapshot_info,
            self.redactions.clone(),
            self.record,
        )
        .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }
}

#[allow(clippy::type_complexity)]
fn wrap_py_fn_snapshot_json(
    py_fn: PyObject,
) -> impl for<'b> Fn(
    &'b Bound<'_, PyTuple>,
    Option<&'b Bound<'_, PyDict>>,
    &'b SnapshotInfo,
    Option<HashMap<String, RedactionType>>,
    bool,
) -> SnapResult<Py<PyAny>>
       + Send
       + Sync {
    move |args: &Bound<'_, PyTuple>,
          kwargs: Option<&Bound<'_, _>>,
          info: &SnapshotInfo,
          redactions: Option<HashMap<String, RedactionType>>,
          record: bool| {
        let py_fn_cloned = Python::with_gil(|py| py_fn.clone_ref(py));

        let call_fn = move |args: &Bound<'_, PyTuple>,
                            kwargs: Option<&Bound<'_, _>>|
              -> SnapResult<PyObject> {
            Python::with_gil(|py| py_fn_cloned.call(py, args, kwargs)).map_err(SnapError::from)
        };

        let wrapped_fn = snapshot_fn_auto_json!(
            call_fn, args, kwargs;
            serialize_macro=assert_json_snapshot_depythonize;
            result_from_str=|content: String| -> SnapResult<PyObject> {
                Python::with_gil(|py| {
                    let value: serde_json::Value = serde_json::from_str(&content)?;
                    let obj = pythonize::pythonize(py, &value).map_err(SnapError::from)?;
                    Ok(obj.into())
                })
            }
        );

        wrapped_fn(args, kwargs, info, redactions, record)
    }
}

#[pyfunction]
pub fn mock_json_snapshot(
    py_fn: PyObject,
    snapshot_info: SnapshotInfo,
    record: bool,
    redactions: Option<HashMap<String, RedactionType>>,
) -> PyResult<PyObject> {
    Python::with_gil(|py| {
        let callable = Py::new(
            py,
            PyMockWrapper {
                f: Box::new(wrap_py_fn_snapshot_json(py_fn)),
                snapshot_info,
                record,
                redactions,
            },
        )?;
        Ok(callable.into())
    })
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use crate::{RedactionType, SnapError, SnapResult};
    use insta::assert_json_snapshot as assert_json_snapshot_macro;
    use insta::internals::SnapshotContents;
    use insta::Snapshot;
    use std::{
        cell::Cell,
        ffi::CString,
        path::{Path, PathBuf},
        rc::Rc,
    };

    use pyo3::{
        types::{PyAnyMethods, PyDict, PyModule, PyTuple},
        Bound, IntoPyObject, Py, PyAny, PyResult, Python,
    };

    use crate::{mock_json_snapshot, SnapshotInfo};

    fn snapshot_folder_path() -> PathBuf {
        // This env var points to the root of your crate during cargo test/build
        let crate_root = Path::new(env!("CARGO_MANIFEST_DIR"));
        crate_root.join("src").join("snapshots")
    }

    #[test]
    fn test_snapshot_json_or_mock_creates_and_reads_snapshot() -> SnapResult<()> {
        let input_1 = 4;

        // Shared counter to track how many times the function is called
        let call_count = Rc::new(Cell::new(0));
        let call_count_clone = Rc::clone(&call_count);

        let f = |i| {
            call_count_clone.set(call_count_clone.get() + 1);
            Ok::<_, SnapError>(i * 2)
        };

        let snaphot_folder_path = snapshot_folder_path();
        let snapshot_info = SnapshotInfo {
            snapshot_folder: snaphot_folder_path,
            snapshot_name: "test_create_snapshot_fn".to_string(),
            relative_test_file_path: None,
            allow_duplicates: true,
        };

        let snapshot_json_or_mock = snapshot_fn_auto_json!(f, x);

        // First run: record mode, should call the function
        let result_1: i32 = snapshot_json_or_mock(input_1, &snapshot_info, None, true)?;
        assert_eq!(result_1, 8);
        assert_eq!(
            call_count.get(),
            1,
            "Function should have been called once during recording"
        );

        // Second run: replay mode, should NOT call the function
        let result_2: i32 = snapshot_json_or_mock(input_1, &snapshot_info, None, false)?;
        assert_eq!(result_2, 8);
        assert_eq!(
            call_count.get(),
            1,
            "Function should NOT have been called again during replay"
        );

        Ok(())
    }

    #[test]
    fn test_create_mocked_pyfn_creates_and_reads_snapshot() -> SnapResult<()> {
        pyo3::prepare_freethreaded_python();
        let snapshot_info = SnapshotInfo {
            snapshot_name: "test_create_mocked_pyfn".to_string(),
            relative_test_file_path: None,
            allow_duplicates: true,
            snapshot_folder: snapshot_folder_path(),
        };

        Python::with_gil(|py| -> PyResult<()> {
            // Define a Python function with a mutable counter
            let code = r#"
counter = {"calls": 0}
def compute(x):
    counter["calls"] += 1
    return {"result": x * 10, "calls": counter["calls"]}
"#;

            let module = PyModule::from_code(
                py,
                CString::new(code)?.as_c_str(),
                CString::new("testmod.py")?.as_c_str(),
                CString::new("testmod")?.as_c_str(),
            )?;
            let py_fn: Py<PyAny> = module.getattr("compute")?.into_pyobject(py)?.into();

            // Wrap with snapshot function in RECORDING mode
            let wrapper_obj =
                mock_json_snapshot(py_fn.clone_ref(py), snapshot_info.clone(), true, None)?;
            let wrapper = wrapper_obj.bind(py);

            let args = PyTuple::new(py, 7.into_pyobject(py))?;

            let result1: Bound<'_, PyDict> = wrapper.call1(args)?.extract()?;
            assert_eq!(result1.get_item("result").unwrap().extract::<i32>()?, 70);
            assert_eq!(result1.get_item("calls").unwrap().extract::<i32>()?, 1);

            let wrapper_obj = mock_json_snapshot(py_fn, snapshot_info.clone(), false, None)?;
            let wrapper = wrapper_obj.bind(py);
            let args = PyTuple::new(py, 7.into_pyobject(py))?;

            let result2: Bound<'_, PyDict> = wrapper.call1(args)?.extract()?;
            assert_eq!(result2.get_item("result").unwrap().extract::<i32>()?, 70);
            assert_eq!(result2.get_item("calls").unwrap().extract::<i32>()?, 1);

            Ok(())
        })?;

        Ok(())
    }
}
