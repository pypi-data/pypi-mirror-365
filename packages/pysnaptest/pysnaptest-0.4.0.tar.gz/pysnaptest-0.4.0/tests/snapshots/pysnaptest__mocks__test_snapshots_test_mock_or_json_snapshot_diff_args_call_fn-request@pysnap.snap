---
source: src/mocks.rs
description: "Test File Path: tests/test_snapshots.py"
---
{
  "args": [
    1
  ],
  "kwargs": {
    "y": 2
  }
}
